### ISSUE TITLE
plugin registration for extensions to endpoints.serverFromString and endpoints.clientFromString

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4695]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4695|
|Type|enhancement|
|Created|2010-10-19 00:57:09Z|
|Branch|https://github.com/twisted/twisted/tree/endpoints-plugins-4695|

Originally [#4473](https://github.com/twisted/twisted/issues/4473) specified a plugin API, but that was biting off more than a single ticket could chew by a wide margin.  This ticket is to complete that work.

As originally specified:

   (`serverFromString` and `clientFromString`) should be implemented in terms of a pluggable API, like the one in `twisted.cred.strcred`, so that third parties can supply their own Twisted transports and built-in protocols can be invoked using them.

<details><summary>Searchable metadata</summary>

```
trac-id__4695 4695
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_endpoints_plugins_4695 branches/endpoints-plugins-4695
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1287449829000000 1287449829000000
changetime__1290481044000000 1290481044000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [30304]) Branching to 'endpoints-plugins-4695'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/endpoints-plugins-4695)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @glyph|
|-|-|

  1. The class docstring for `IStreamClientEndpointStringParser` ends with "plugin method" but probably means "prefix method"
  1. Why is prefix a method and not just an attribute?
  1. The unrecognized prefix case of `clientFromString` is untested and probably should more closely resemble the behavior of `serverFromString` in that case - currently the caller will get _TypeError: 'NoneType' object is not callable_.
  1. Needs a news fragment

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [30338]) plugin -> prefix re [#4695](https://github.com/twisted/twisted/issues/4695) point 1

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

(In [30339]) re [#4695](https://github.com/twisted/twisted/issues/4695) review point 4: NEWS file

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

(In [30340]) re [#4695](https://github.com/twisted/twisted/issues/4695) review point 2: make prefix an attribute instead of a method

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [30342]) re [#4695](https://github.com/twisted/twisted/issues/4695) review point 3: failing test for unrecognized client prefix

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [30343]) re [#4695](https://github.com/twisted/twisted/issues/4695) review point 3: make the test for unrecognized prefix pass.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph removed owner|
|-|-|

Build results (see link above) ready with the usual potpourri of Windows failures; I think that addresses all the points of feedback.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @glyph|
|-|-|

  1. The first line of endpoints.py belongs somewhere else

Otherwise, great, please merge!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set status to `closed`|
|-|-|

(In [30348]) Merge endpoints-plugins-4695.

Author: glyph

Reviewer: exarkun

Fixes: [#4695](https://github.com/twisted/twisted/issues/4695)

Add plugin interfaces to allow third parties to implement their own client and
server endpoint types, and have them be parsed by clientFromString,
serverFromString, and strports.service().

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
